tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,752 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Head chunk file header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the beginning of a head chunk file file.
+	MagicHeadChunks = 0x0130BC91
+
+	headChunksFormatV1 = 1
+	inBufferShards     = 128             // 128 is a randomly chosen number.
+	writeBufferSize    = 2 * 1024 * 1024 // 2 MiB.
+)
+
+var (
+	// DefaultHeadChunkFileMaxTimeRange is the default head chunk file time range.
+	// Assuming a general scrape interval of 15s, a chunk with 120 samples would
+	// be cut every 30m, so anything <30m will cause lots of empty files. And keeping
+	// it exactly 30m also has a chance of having empty files as its near that border.
+	// Hence keeping it a little more than 30m, i.e. 40m.
+	DefaultHeadChunkFileMaxTimeRange = 40 * int64(time.Minute/time.Millisecond)
+
+	// ErrHeadReadWriterClosed returned by any method indicates
+	// that the HeadReadWriter was closed.
+	ErrHeadReadWriterClosed = errors.New("HeadReadWriter closed")
+)
+
+// CorruptionErr is an error that's returned when corruption is encountered.
+type CorruptionErr struct {
+	Dir       string
+	FileIndex int
+	Err       error
+}
+
+func (e *CorruptionErr) Error() string {
+	return fmt.Sprintf("corruption in head chunk file %s: %s", segmentFile(e.Dir, e.FileIndex), e.Err.Error())
+}
+
+// ChunkDiskMapper is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type ChunkDiskMapper struct {
+	// Writer.
+	dirFile *os.File
+	curFile *os.File // File being written to.
+
+	curFileMint     int64 // In milliseconds.
+	curFileMaxt     int64 // In milliseconds.
+	curFileSequence int
+	n               int64 // Bytes written in current open file.
+	maxFileDuration int64 // Duration in milliseconds.
+
+	buf                 [8]byte
+	wbuf                *bufio.Writer
+	wbufLock            sync.RWMutex
+	crc32               hash.Hash
+	inBufferChunks      [inBufferShards]map[uint64]chunkenc.Chunk
+	inBufferChunksLocks [inBufferShards]sync.RWMutex
+
+	// Reader.
+	// The int key in the map is the file number on the disk.
+	mmappedChunks    map[int]ByteSlice
```

Comment:
Those are not chunks.. but chunkFiles? 
```suggestion
    mmappedChunksFiles    map[int]ByteSlice
```